### PR TITLE
Release automation

### PR DESCRIPTION
**Pull request by nventuro**
Fixes https://github.com/OpenZeppelin/openzeppelin-solidity/issues/157

We could write the changelog update script in bash and get rid of `moment`, but this does work as-is.

### CODE REVIEW

File: scripts/release/release.sh [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+# Exit script as soon as a command fails.
+set -o errexit
+
+log() {
+  # Print to stderr to prevent this from being 'returned'
+  echo "$@" > /dev/stderr
+}
+
+current_version() {
+  echo v$(node --print --eval "require('./package.json').version")
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
  echo "v$(node --print --eval "require('./package.json').version")"
```

File: scripts/release/release.sh [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+current_release_branch() {
+  v=$(current_version)
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
  v="$(current_version)"
```

File: scripts/release/release.sh [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  echo "release-${v%%-rc.*}"
+}
+
+assert_current_branch() {
+  current_branch=$(git symbolic-ref --short HEAD)
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
  current_branch="$(git symbolic-ref --short HEAD)"
```

File: scripts/release/release.sh [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  expected_branch=$1
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
  expected_branch="$1"
```

File: scripts/release/release.sh [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  if [[ "$current_branch" != "$expected_branch" ]]; then
+    log "Current branch '$current_branch' is not '$expected_branch'"
+    exit 1
+  fi
+}
+
+push_release_branch_and_tag() {
+  git push upstream $(current_release_branch) $(current_version)
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
  git push upstream "$(current_release_branch)" "$(current_version)"
```

File: scripts/release/release.sh [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+push_and_publish() {
+  dist_tag=$1
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
  dist_tag="$1"
```

File: scripts/release/release.sh [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+  log "Pushing release branch and tags to upstream"
+  push_release_branch_and_tag
+
+  log "Publishing package on npm"
+  npm publish --tag $dist_tag --otp $(prompt_otp)
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
  npm publish --tag "$dist_tag" --otp "$(prompt_otp)"
```

File: scripts/release/release.sh [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+  if [[ "$dist_tag" == "latest" ]]; then
+    npm dist-tag rm --otp $(prompt_otp) openzeppelin-solidity next
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
    npm dist-tag rm --otp "$(prompt_otp)" openzeppelin-solidity next
```

File: scripts/release/release.sh [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  fi
+}
+
+prompt_otp() {
+  log -n "Enter npm 2FA token: "
+  read otp
+  echo $otp
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
  echo "$otp"
```

File: scripts/release/release.sh [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+environment_check() {
+  if [[ "upstream" != *"$(git remote)"* ]]; then
```
Comment (frangio) [inline scripts/release/release.sh]: Have you tested this line? It doesn't seem to work as you intended. Here's an alternative:

```suggestion
  if ! git remote get-url upstream &> /dev/null; then
```

File: scripts/release/release.sh [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    log "Will publish as '$(npm whoami)'"
+  else
+    log "Not logged in into npm, run 'npm login' first"
+    exit 1
+  fi
+}
+
+environment_check
+
+if [[ "$@" == "start minor" ]]; then
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
if [[ "$*" == "start minor" ]]; then
```

`"$@"` expands to several strings, one for each element in the `$@` array. This seems to work with `[[` and `==`  but I'd rather change it to `"$*"` which is actually one string.

File: scripts/release/release.sh [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+  assert_current_branch master
+
+  # Create temporary release branch
+  git checkout -b release-temp
+
+  # This bumps minor and adds rc suffix, commits the changes, and tags the commit
+  npm version preminor --preid=rc
+
+  # Rename the release branch
+  git branch --move $(current_release_branch)
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
  git branch --move "$(current_release_branch)"
```

File: scripts/release/release.sh [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+
+  push_and_publish next
+
+elif [[ "$@" == "rc" ]]; then
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
elif [[ "$*" == "rc" ]]; then
```

File: scripts/release/release.sh [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+  log "Bumping pre-release"
+
+  assert_current_branch $(current_release_branch)
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
  assert_current_branch "$(current_release_branch)"
```

File: scripts/release/release.sh [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+  # Bumps rc number, commits and tags
+  npm version prelease
+
+  push_and_publish next
+
+elif [[ "$@" == "final" ]]; then
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
elif [[ "$*" == "final" ]]; then
```

File: scripts/release/release.sh [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+  # Update changelog release date, remove rc suffix, tag, push to git, publish in npm, remove next dist-tag
+  log "Creating final release"
+
+  assert_current_branch $(current_release_branch)
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
  assert_current_branch "$(current_release_branch)"
```

File: scripts/release/release.sh [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+  npm version patch
+
+  push_release_branch_and_tag
+
+  push_and_publish latest
+
+  log "Remember to merge the release branch into master and push upstream"
+
+else
+  log "Unknown command: '$@'"
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
  log "Unknown command: '$*'"
```

File: scripts/release/update-changelog-release-date.js [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+const moment = require('moment');
+
+const pkg = require('../../package.json');
+if (pkg.version.indexOf('-rc') !== -1) {
+  process.exit(0);
+}
+
+const version = pkg.version.replace(/-.*/, ''); // Remove the rc suffix
+
+const changelog = fs.readFileSync('CHANGELOG.md').toString();
```
Comment (frangio) [inline scripts/release/update-changelog-release-date.js]:
```suggestion
const changelog = fs.readFileSync('CHANGELOG.md', 'utf8');
```

File: scripts/release/update-changelog-release-date.js [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+// We need to add the date in a YYYY-MM-DD format, so that it looks like this:
+// ## 2.5.3 (2019-04-25)
+
+if (changelog.indexOf(`## ${version} (unreleased)`) === -1) {
+  throw Error(`Found no changelog entry for version ${version}`);
+}
+
+fs.writeFileSync('CHANGELOG.md', changelog.replace(
+  `## ${version} (unreleased)`,
+  `## ${version} (${moment().format('YYYY-MM-DD')})`)
```
Comment (frangio) [inline scripts/release/update-changelog-release-date.js]: We can remove the moment dependency.

```suggestion
  `## ${version} (${new Date().toISOString().split('T')[0]})`)
```

File: scripts/release/update-changelog-release-date.js [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+);
+
+cp.execSync('git add CHANGELOG.md');
```
Comment (frangio) [inline scripts/release/update-changelog-release-date.js]: I suspect we don't want to ignore the output.

```suggestion
cp.execSync('git add CHANGELOG.md', { stdio: 'inherit' });
```

File: scripts/release/release.sh [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+  npm publish --tag "$dist_tag" --otp "$(prompt_otp)"
+
+  if [[ "$dist_tag" == "latest" ]]; then
+    npm dist-tag rm --otp $(prompt_otp) openzeppelin-solidity next
+  fi
+}
+
+prompt_otp() {
+  log -n "Enter npm 2FA token: "
+  read otp
```
Comment (frangio) [inline scripts/release/release.sh]:
```suggestion
  read -r otp
```

> -r	do not allow backslashes to escape any characters

This flag is usually recommended.

File: scripts/release/release.sh (hunk 10 above)
Comment (nventuro) [reply scripts/release/release.sh]: I did test it, it checks if there is a remote named `upstream`. Yours is clearer though.

Comment (frangio) [reply scripts/release/release.sh]: I tested it and it didn't work for me. Might have been a coincidence that it worked in your test environment.

**Pull request merged**